#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv ('/content/drive/MyDrive/Colab Notebooks/Data/supermarket_sales.csv')

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [4]:
df.groupby (by = 'Branch').Total.sum().round (2)

Branch
A    106200.37
B    106197.67
C    110568.71
Name: Total, dtype: float64

Подивившись сумарний прибуток, бачимо, що найприбутковішим є філіал 'C' - 110 568,71

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [5]:
df[df ['Branch'] == 'C']['City'].nunique('City')
# перевірила чи філіал з найвищим прибутком розташовується в одному місці, чи може в декількох

1

In [6]:
# перевірка показала, що в нас лише одне унікальне значення і можна вивести лише перший рядок для розуміння міста
df[df ['Branch'] == 'C']['City'].head (1)

1    Naypyitaw
Name: City, dtype: object

In [7]:
# для розумінння потрібен чи ні ще один магазин
# по-перше подививось статистику по всім містам
df.groupby (by = 'City').Total.agg(['count','sum', 'mean', 'min', 'max']).round(3)

count         sum     mean     min      max
City                                                  
Mandalay     332  106197.672  319.873  18.638  1022.49
Naypyitaw    328  110568.706  337.100  10.678  1042.65
Yangon       340  106200.370  312.354  12.694  1039.29

In [8]:
# подивимось розподіл продажів по категоріям і містах
goods_in_city = pd.pivot_table ( df, index= 'Product line', columns= 'City', values='Invoice ID', aggfunc='count')
goods_in_city

City                    Mandalay  Naypyitaw  Yangon
Product line                                       
Electronic accessories        55         55      60
Fashion accessories           62         65      51
Food and beverages            50         66      58
Health and beauty             53         52      47
Home and lifestyle            50         45      65
Sports and travel             62         45      59

In [9]:
# середній рейтінг по всім містам
df.groupby (by = 'City').Rating.agg(['mean'])

mean
City               
Mandalay   6.818072
Naypyitaw  7.072866
Yangon     7.027059

In [10]:
# розподіл по типам кліентів по містах
cust = pd.pivot_table ( df, index= 'City', columns= 'Customer type', values='Invoice ID', aggfunc='count')
cust

Customer type  Member  Normal
City                         
Mandalay          165     167
Naypyitaw         169     159
Yangon            167     173

**відповідь**

Зробивши невеликий аналіз, бачимо, що в **Naypyitaw** найвищий прибуток, середній чек та max чек у порівняні з іншими містами. Найбільше купують товари категорії **Fashion accessories (65)** та **Food and beverages (66)**, а найменше  - **Home and lifestyle (45)** та **Sports and travel (45**).

Середній рейтінг магазинів в цьому місті **7,07**  - найвищий серед усіх.

Розподіл між членами магазину і звичийними покупцями на користь членів 169 до 159. Також можна відмітити, що в цьому місті найбільша кількість членів.

**висновок**

Магазин є сенс відкрити,так як залученість покупців висока (рейтінг, членство, високий середній і максимальний чек). Проте треба переглянути асортимент і зробити акцент на популярні товари для залученності більшої кількості покупців

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [11]:
# щоб мати таблицю перед очима, з назвами стовпців, видами даних для більш комфортної роботи
df.head (1)

Invoice ID Branch    City Customer type  Gender       Product line  \
0  750-67-8428      A  Yangon        Member  Female  Health and beauty   

   Unit price  Quantity   Tax 5%     Total      Date   Time  Payment    cogs  \
0       74.69         7  26.1415  548.9715  1/5/2019  13:08  Ewallet  522.83   

   gross margin percentage  gross income  Rating  
0                 4.761905       26.1415     9.1

In [12]:
invoices_by_category = pd.pivot_table ( df, index= 'Branch', columns= 'Product line', values='Invoice ID', aggfunc='count')

In [13]:
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [14]:
max_in_electr = invoices_by_category['Electronic accessories'].max()
max_electr_in_branch = invoices_by_category['Electronic accessories'].idxmax()
print (f'Найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" в філіалі {max_electr_in_branch} і кількість дорівнює {max_in_electr}')

Найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" в філіалі A і кількість дорівнює 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**Гіпотези**

так як ми вже рахували загальний дохід по містам, філіалам, валовий прибуток і перевіряли, що маржа однакова для всих магазинів в різних містах

Вважаю доцільним ще дослідити:

1. розподіл між звичайними покупцями та членами магазину в різних містах (для розуміння треба робити якісь дії для залучення кількості членів магазинів)

2. розподіл товарів по містам (додавання або видалення категорій, або перерозподіл)

3. використання системи оплати (аналіз та залучення інших методів)

4. проаналізувати прибуток по місяцям (коли робити додаткові дії для залучення клиентів)


частково це дослідила для відповіді на питання 2, тому трошки будуть повтори)))

In [15]:
# розподіл по типам кліентів по містах
custom = pd.pivot_table ( df, index= 'Customer type', columns= 'City', values='Invoice ID', aggfunc='count')

In [16]:
custom

City           Mandalay  Naypyitaw  Yangon
Customer type                             
Member              165        169     167
Normal              167        159     173

In [17]:
# той же розподіл в %
perc_custom = custom.div(custom.sum(axis=0), axis=1) * 100

In [18]:
perc_custom.round (2)

City           Mandalay  Naypyitaw  Yangon
Customer type                             
Member             49.7      51.52   49.12
Normal             50.3      48.48   50.88

In [19]:
concatenated_df = pd.concat([custom, perc_custom], keys=['Кількість', 'Доля %'], axis=1)

In [20]:
concatenated_df.round (2)

Кількість                    Доля %                 
City           Mandalay Naypyitaw Yangon Mandalay Naypyitaw Yangon
Customer type                                                     
Member              165       169    167     49.7     51.52  49.12
Normal              167       159    173     50.3     48.48  50.88

In [21]:
concatenated_df.loc['Total'] = concatenated_df.sum(numeric_only=True).values

хотіла спробувати, як пояснювала Ганна - хитрий pivot table)))

In [22]:
concatenated_df.round (2)

Кількість                    Доля %                  
City           Mandalay Naypyitaw Yangon Mandalay Naypyitaw  Yangon
Customer type                                                      
Member            165.0     169.0  167.0     49.7     51.52   49.12
Normal            167.0     159.0  173.0     50.3     48.48   50.88
Total             332.0     328.0  340.0    100.0    100.00  100.00

In [23]:
# подивимось розподіл продажів категорій товарів по містах
goods_in_city = pd.pivot_table ( df, index= 'Product line', columns= 'City', values='Invoice ID', aggfunc='count')
goods_in_city

City                    Mandalay  Naypyitaw  Yangon
Product line                                       
Electronic accessories        55         55      60
Fashion accessories           62         65      51
Food and beverages            50         66      58
Health and beauty             53         52      47
Home and lifestyle            50         45      65
Sports and travel             62         45      59

In [24]:
# цей же розподіл, але у відсотках по кожній категорії товарів
perc_goods = goods_in_city.div(goods_in_city.sum(axis=0), axis=1) * 100

In [25]:
perc_goods.round (2)

City                    Mandalay  Naypyitaw  Yangon
Product line                                       
Electronic accessories     16.57      16.77   17.65
Fashion accessories        18.67      19.82   15.00
Food and beverages         15.06      20.12   17.06
Health and beauty          15.96      15.85   13.82
Home and lifestyle         15.06      13.72   19.12
Sports and travel          18.67      13.72   17.35

In [26]:
# знайдемо категорію, яку найбільше і найменше купують в кожному місті
for city in goods_in_city.columns:
    max_category = goods_in_city[city].idxmax()
    min_category = goods_in_city[city].idxmin()
    max_goods = goods_in_city [city].max()
    min_goods = goods_in_city [city].min()
    max_perc = round ((perc_goods [city].max()),2)
    min_perc = round ((perc_goods [city].min()),2)
    print(f"{city}:")
    print(f" Найбільше продажів: {max_category} {max_goods} і це {max_perc} %")
    print(f" Найменше продажів: {min_category} {min_goods}  і це {min_perc} %\n")

Mandalay:
 Найбільше продажів: Fashion accessories 62 і це 18.67 %
 Найменше продажів: Food and beverages 50  і це 15.06 %

Naypyitaw:
 Найбільше продажів: Food and beverages 66 і це 20.12 %
 Найменше продажів: Home and lifestyle 45  і це 13.72 %

Yangon:
 Найбільше продажів: Home and lifestyle 65 і це 19.12 %
 Найменше продажів: Health and beauty 47  і це 13.82 %



In [27]:
# подивимось розподіл використання систем оплат по містам
view_of_payment = pd.pivot_table ( df, index= 'City', columns= 'Payment', values='Invoice ID', aggfunc='count')
view_of_payment.loc ['Total'] = view_of_payment.sum ()
view_of_payment

Payment    Cash  Credit card  Ewallet
City                                 
Mandalay    110          109      113
Naypyitaw   124           98      106
Yangon      110          104      126
Total       344          311      345

In [28]:
# знайдемо місяць із дати
df ['Month'] = pd.to_datetime(df['Date']).dt.month

In [29]:
df.head(2)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   

             Product line  Unit price  Quantity   Tax 5%     Total      Date  \
0       Health and beauty       74.69         7  26.1415  548.9715  1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200  3/8/2019   

    Time  Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08  Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29     Cash   76.40                 4.761905        3.8200     9.6   

   Month  
0      1  
1      3

In [30]:
# згруппуємо дохід по місяцям
df.groupby (by ='Month').Total.sum().round (2)

Month
1    116291.87
2     97219.37
3    109455.51
Name: Total, dtype: float64

In [31]:
# проаналізуемо дохід по місяцям і містам (вивела в більш зручному форматі для читання данних (для себе...більш зручному))
total_month_city = pd.pivot_table (df, index= 'Month', columns= 'City', values= 'Total', aggfunc= 'sum')
total_month_city.loc['Total'] = total_month_city.sum()
total_month_city ['Total'] = total_month_city.sum (axis = 1)
#total_month_city.round (2)
total_month_city.style.format("{:,.0f}")

In [32]:
 # перевірка, що табл зверху рахується вірно
 total_table = df['Total'].sum()
 total_table

np.float64(322966.749)

1. Подивившись на розподіл між звичайними покупцями і тими, хто є членами магазину, бачимо, що в місті **Naypyitaw** найбільше покупців з картками (169) і в той же час найменше без карток (159). При цьому в **Yangon** найбільше людей без карток (173). Є сенс попрацювати з лояльністю покупців і збільшувати кількість членів в усіх містах, так як розподіл між членами і звичайними покупцями майже 50 на 50 в усіх містах...

2. по містах і товарах маємо таку інформацію:

    **Mandalay:**

   Найбільше продажів: Fashion accessories 62 і це 18.67 %

   Найменше продажів: Food and beverages 50  і це 15.06 %

    **Naypyitaw:**

  Найбільше продажів: Food and beverages 66 і це 20.12 %

  Найменше продажів: Home and lifestyle 45  і це 13.72 %

    **Yangon:**

  Найбільше продажів: Home and lifestyle 65 і це 19.12 %

  Найменше продажів: Health and beauty 47  і це 13.82 %

  необхідно більш детально проаналізувати продаж товарів в кожному магазині, зменшити асортимент товарів, які купують мало, також додати аналітику по прибутковості кожної категориї, як варіант провести маркетингові дії для залучення покупців на категорії, де менше продажів

3. Проаналізувавши розподіл оплат, бачимо, що найменше використовують кредітні картки (тут треба більш детально уточнити взагалі картки, чи дійсно кредитні), **cash** i **ewallet** використовують однаково в загаліній статистиці.
По містах: **cash** більше використовують в **Naypyitaw**, **ewallet** в **Yangon**
Один з варіантів збільшення використання карток і електронного гаманця, зручніше оплата (каси самообслуговування), зменшення витрат на персонал, комісії банків

4. Проаналізувавши дохід по місяцям, бачимо, що найбільший дохід в січні (116 291.87), найменший в лютому (97219.37). В розбивці по містам ситуація не змінилась, найменший показник продажів в лютому. Одна з причин менше днів в місяці, інша - немає свят і це природній спад. В цей період можна додавати маркетингові дії для збільшення продажів. Для більш детальної аналітики і виявлення патернів бажано порівняти з іншими роками в той же період.

***Загальні висновки - рекомендації:***
1. залучати людей мати картки членів магазину (більше ймовірність, що прийдуть в ці магазини)
2. Звернути увагу на розподіл продажі товарів, як варіант вивести з асортименту ті, шо менш продаються, або додати маркетингових дій до залучення їх придбання (більш детальна аналітика по прибутковості кожного товару)
3. спонукати розраховуватись картками/ел. гаманцем (зменшення витрат на "обслуговування" готівки)
4. в місяцях, де немає свят, менше днів, вихідних - стимулювати продаж, щоб не було провалів по доходу